File: perfect2.py
#!/usr/bin/env python3

##                                     ##
## program that finds perfect numbers  ##
##                                     ##

from math import sqrt

print("")
print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
print(" This program finds perfect numbers using Mersenne Primes ")
print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")
print("")

# Lucas-Lehmer prime test for odd p > 2
def LLT(p):
    s = 4
    M = (1<<p) - 1
    for n in range(0,p-1):
        s = ((s * s) - 2) % M;
        if(s == 0):
            return False
    return True

print("The perfect numbers:")
print(6)

# find the perfect numbers
p = 1
while True:
    p += 2

#   LLT check
    if LLT(p):
        continue

#   summation var of divisors
    psum = 1

#   the potential perfect number
####perfect = 2**(p-1)*(2**(p)-1)
    perfect = (1<<(2*p-1))-(1<<(p-1))

#   the limit to search to
    sqrtp = int(sqrt(perfect))

#   add up all the divisors into psum
    for n in range(2,sqrtp+1):
        if perfect % n == 0:
            psum += n
            psum += perfect // n

#   get rid of possible extra summation
    if sqrtp**2 == perfect:
        psum -= sqrtp

#   if psum is equal to the potenial perfect number, we have a match
    if psum == perfect:
        print(perfect)

File: perfect2.dart
#!/usr/bin/env dart

/*
 * program that finds perfect numbers
*/

import 'dart:math';
import 'dart:io';

// Lucas-Lehmer prime test for odd p > 2
bool LLT(int p) {
        BigInt s = BigInt.two + BigInt.two;
        BigInt M = BigInt.two.pow(p) - BigInt.one;
        for(int n = 0; n < p-2; n += 1) {
                s = ((s * s) - BigInt.two) % M;
                if(s == BigInt.zero) {
                        return false;
                }
        }
        return true;
}

// Newton's method for square roots
BigInt NewtonSqrt(BigInt perfect) {
        double? root,check;
        double x = perfect / BigInt.two;
        double n = x;
        while(true) {
                root = 0.5 * (x + (n / x));
                check = root - x;
                if(check > 0 && check < 1) {
                        break;
                }
                else if(check < 0 && check > -1) {
                        break;
                }
                x = root;
        }
        return (BigInt.from(root) + BigInt.two); 
}

void main() {
        print("");
        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
        print(" This program finds perfect numbers using Mersenne Primes ");
        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");
        print("");

        print("The perfect numbers:");
        print(6);

        // find the perfect numbers
        for(int p = 3; p < 21; p+=2) {

                // LLT test
                if(LLT(p)) {
                        continue;
                }

                BigInt psum = BigInt.one;

                // the potential perfect number
                BigInt perfect = BigInt.two.pow(p-1)*(BigInt.two.pow(p) - BigInt.one);

                // use Newton's method for sqrt for stopping point
                BigInt stop = NewtonSqrt(perfect);

                // add up all the divisors into psum
                for(BigInt n = BigInt.two; n < stop; n = n + BigInt.one) {
                        if(perfect % n == BigInt.zero) {
                                psum = psum + n;
                                psum = psum + BigInt.from(perfect / n);
                        }
                }

                // get rid of extra possible summation
                if(stop*stop == perfect) {
                        psum = psum - stop;
                }


                // if psum is equal to the potenial perfect number, we have a match
                if(psum == perfect) {
                        print(perfect);
                }
        }
}
